### DISCUSSION TITLE
Include Code CIDs in Events

### DISCUSSION THREAD

**Discussion by Stebalien** [filecoin-project/FIPs#753] [Filecoin Virtual Machine]
At the moment, events don't include code CIDs, just the emitter's ID. Unfortunately:

- It's entirely possible for an actor to be ephemeral (exist and self-destruct within a single epoch).
- It _will_ (very likely) be possible for actors to upgrade their own code during an epoch (see #396).

Neither of these are an issue today as:

1. Only the EVM emits events (for now).
2. The EVM doesn't upgrade and/or self destruct (at least it doesn't _fully_ self destruct).

However, it may be an issue in the future. Even now, I'd like a more convenient way to say "give me all events from actors with code CIDs X with some label Y".

The major _downside_ of this is that events get larger. On the other hand, the _cost_ of said bytes should be negligible as events aren't a part of the state (and are only persisted for indexing/historical purposes).

**Comment by anorth** [comment]
How relevant are these considerations in other ecosystems, like Ethereum? Contracts on Ethereum can also self-destruct. While they don't directly upgrade, proxy patterns etc achieve something quite similar.

Does this cause problems for Ethereum ecosystem devs? Can we find discussions where they wish they did it differently?

Note: the Filecoin EVM actors _will_ upgrade, but will do so silently as part of a network upgrade unless something is changed about how built-in actors are deployed.

**Reply by Stebalien** [reply]
I'm not sure.

It's slightly less of an issue because self-destructs happen at the _end_ of an epoch, not immediately. There's also a self-destruct/redeploy pattern of upgrades, but again, because the self-destruct part happens at the end of the epoch, you can't have multiple actor codes with the same addresses emit events during the same epoch/message.

Really, there are three users:

- Dapp developers probably just care about events emitted by specific contracts. E.g., your wallet, an ERC20 token, etc.
- Wallet (e.g., metamask) developers need to know about new ERC20 token deposits, but they likely just look for all events that "look" like ERC20 token deposits. In this case, looking at code CIDs is likely detrimental as code CIDs will change...
- Chain explorers generally don't care about "ephemeral" contracts, but likely do care about code CIDs.

So I think the user that might have difficulty is that last one: chain explorers. In Ethereum, due the restrictions on self=destruct, if an actor has some non-empty bytecode at the beginning/end of an epoch, it didn't have any other non-empty bytecode during the epoch. This may _not_ be true for Filecoin which could introduce some... interesting challenges for explorers.

On the other hand... for explorers, we might be better off just augmenting execution traces with:

1. Events as they're emitted.
2. "Upgrade" events so they can keep track of code CID changes.

So maybe this is all just a non-issue.

**Comment by Stebalien** [comment]
I'm going to mark this as rejected as we've sat on this for a while and don't have a strong motivation for making it happen.
